File: 1_csharp/Aoc.cs
using System;
using System.Linq;
using System.Collections.Generic;
using System.Collections.Immutable;

namespace aoc
{
  class Aoc
  {
    public static int FuelRequired(int mass) => (mass / 3) - 2;
    public static int TotalFuelRequired(int mass, int total)
    {
      int f = FuelRequired(mass);
      if (f <= 0)
        return total;
      else
        return TotalFuelRequired(f, total + f);
    }

    public static int PartOne(string input) =>
      Parse(input).Select(i => FuelRequired(i)).Sum();

    public static int PartTwo(string input) => 
      Parse(input).Select(i => TotalFuelRequired(i, 0)).Sum();

    private static IEnumerable<int> Parse(string input) =>
      input.Trim().Split("\n").Select(s => int.Parse(s));
  }
}

File: 1_elixir/lib/aoc.ex
defmodule Aoc do
  def fuel_required(mass),
    do: (div(mass, 3) - 2) |> max(0)

  def total_fuel_required(mass),
    do: total_fuel_required(mass, 0)

  def total_fuel_required(0, total),
    do: total

  def total_fuel_required(mass, total) do
    f = fuel_required(mass)
    total_fuel_required(f, total + f)
  end

  @spec part_one(binary) :: number
  def part_one(input),
    do:
      input
      |> parse
      |> Enum.map(&fuel_required/1)
      |> Enum.sum()

  def part_two(input),
    do:
      input
      |> parse
      |> Enum.map(&total_fuel_required/1)
      |> Enum.sum()

  def parse(input),
    do:
      input
      |> String.trim()
      |> String.split()
      |> Enum.map(&String.to_integer/1)
end
